src/osd/PeeringState.cc

```diff
@@ -6453,6 +6466,7 @@ boost::statechart::result PeeringState::GetInfo::react(const QueryState& q)
 void PeeringState::GetInfo::exit()
 {
   context< PeeringMachine >().log_exit(state_name, enter_time);
+  ceph_assert(peer_info_requested.empty());
```

Comment:
Is this actually true in general?  What if we get an interval change before GetInfo completes?

Reply:
After rereading I see a path leading to assertion failure. If `prior_set.pg_down` is set, the `probe` container **can still** enlist shards that primary should to send `pg_query_t::INFO` to. Therefore, they can be present in the `peer_info_requested` when `exit()` is called in the effect of dispatching `IsDown` from the constructor.

When `GetInfo` handles  `MNotifyRec` that affects `last_epoch_started`, `peer_info_requested` is cleared from the OSDs that aren't in the probe set anymore.

How about:

```cpp
  ceph_assert(peer_info_requested.empty() || prior_set.pg_down);
```
?

Reply:
I still don't think that's right.  If Started (parent of GetInfo) gets an AdvMap event for which should_restart_peering is true, it will transition directly to Reset.  The inner states (Primary/Peering/GetInfo) here would each have ::exit() and then their destructor called.  In that case, peer_info_requested may well be non-empty, and that's perfectly ok.  peer_info_requested is a member of GetInfo, and will cease to exist.  When we next enter GetInfo, it'll be a new and empty map.

Reply:
Thanks! I will drop the commit.